torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+        if node.target in reduction_ops and len(shape) == 0:
+            operand = node.all_input_nodes[0]
+            if operand in self.batch_dim_map:
+                operand_batch_dim = self.get_batch_dim(operand)
+                if operand_batch_dim == 0:
+                    reduction_over_batch = True
+                    self.set_batch_dim(node, operand_batch_dim)
+
+            else:
+                raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        elif input_full_reduction:
+            # the first consumer node that consumes the full reduction
+            operand = node.all_input_nodes[0]
+            assert (
+                operand in self.batch_dim_map
+            ), "input have full reduction but not in dim map!"
+            self.set_batch_dim(node, self.get_batch_dim(operand))
+        else:
+            for i, dim_size in enumerate(shape):
+                if dim_size == self.batch_dim_size:
+                    self.set_batch_dim(node, i)
+
+        node_batch_dim = self.get_batch_dim(node)
+        batch_dim_shard_spec = DTensorSpec(
+            mesh=mesh, placements=[Shard(node_batch_dim)]
+        )
+        if reduction_over_batch or input_full_reduction:
+            batch_dim_shard_spec.from_local = True  # type: ignore[attr-defined]
+
+        return batch_dim_shard_spec, reduction_over_batch
+
+
+def _gen_shard_strategy(
+    mesh: DeviceMesh, shard_dim: int, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a shard strategy on shard_dim
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Shard(shard_dim)]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_replicate_strategy(
+    mesh: DeviceMesh, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a replicate strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Replicate()]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
+    """
+    util function to generate a partial strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[_Partial()]),
+    )
+
+
+def build_data_parallel_strategies(
+    train_step_graph: GraphModule,
+    num_params: int,
+    num_states: int,
+    mesh: DeviceMesh,
+    batch_dim: int = 0,
+) -> Dict[fx.Node, StrategyList]:
+    """
+    This function loop through the train step graph and build the
+    data parallel strategy for each fx Node
+    """
+    activation_idx = num_params + num_states
+    non_compute_ops = [
+        torch.ops._spmd.tag_grad.default,
+        aten.ones_like.default,
+        aten.t.default,
+        aten.view.default,
+        aten.reshape.default,
+        aten.detach.default,
+        aten.clone.default,
```

Comment:
What makes something non-compute? I see both view and non-view operations in this list. Is this anything that's bandwidth bound?

Reply:
I guess, most of these make sense except `ones_like`, that's surprising.

Reply:
Mostly a non-compute ops does not really do computation and only do some simple transformations (view, reshape, etc). yeah let me re-think `ones_like` and how to handle it